### PR TITLE
[libc++][TZDB] Finishes zoned_time constructors.

### PR DESCRIPTION
**Pull request by mordante**
Completes
- LWG3225 zoned_time converting constructor shall not be noexcept
- LWG3226 zoned_time constructor from string_view should accept zoned_time<Duration2, TimeZonePtr2>

Implements parts of:
- P0355 Extending to chrono Calendars and Time Zones

### CODE REVIEW

File: libcxx/include/__chrono/zoned_time.h [hunk 1]
```diff
@@ -76,12 +81,71 @@ class zoned_time {
...
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(_TimeZonePtr __zone, const local_time<_Duration>& __tp)
+    requires(is_convertible_v<decltype(std::declval<_TimeZonePtr&>() -> to_sys(local_time<_Duration>{})),
+                              sys_time<duration>>)
+      : __zone_{std::move(__zone)}, __tp_{__zone_->to_sys(__tp)} {}
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(string_view __name, const local_time<_Duration>& __tp)
+    requires(requires {
+      _TimeZonePtr{__traits::locate_zone(string_view{})};
+    } && is_convertible_v<decltype(std::declval<_TimeZonePtr&>() -> to_sys(local_time<_Duration>{})),
```
Comment (ldionne) [inline libcxx/include/__chrono/zoned_time.h:107]:
```suggestion
    requires is_constructible<zoned_time, decltype(__traits::locate_zone(__name))> && is_convertible_v<decltype(std::declval<_TimeZonePtr&>() -> to_sys(local_time<_Duration>{})),
```

The constraint in http://eel.is/c++draft/time.zone#zonedtime.ctor-18.sentence-1 is:
> Constraints: zoned_time is constructible from the return type of `traits​::​locate_zone(name)` and `tp`.

File: libcxx/include/__chrono/zoned_time.h [hunk 2]
```diff
@@ -76,12 +81,71 @@ class zoned_time {
...
+  template <class _Duration2>
+  _LIBCPP_HIDE_FROM_ABI zoned_time(const zoned_time<_Duration2, _TimeZonePtr>& __zt)
+    requires is_convertible_v<sys_time<_Duration2>, sys_time<_Duration>>
+      : __zone_{__zt.get_time_zone()}, __tp_{__zt.get_sys_time()} {}
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(_TimeZonePtr __zone, const sys_time<_Duration>& __tp)
+      : __zone_{std::move(__zone)}, __tp_{__tp} {}
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(string_view __name, const sys_time<_Duration>& __tp)
+    requires requires { _TimeZonePtr{__traits::locate_zone(string_view{})}; }
```
Comment (ldionne) [inline libcxx/include/__chrono/zoned_time.h:96]: Similar comment about the constraint here.

File: libcxx/include/__chrono/zoned_time.h [hunk 3]
```diff
@@ -76,12 +81,71 @@ class zoned_time {
...
+      : zoned_time{__traits::locate_zone(__name), __tp} {}
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(_TimeZonePtr __zone, const local_time<_Duration>& __tp, choose __c)
+    requires(is_convertible_v<
+                decltype(std::declval<_TimeZonePtr&>() -> to_sys(local_time<_Duration>{}, choose::earliest)),
+                sys_time<duration>>)
+      : __zone_{std::move(__zone)}, __tp_{__zone_->to_sys(__tp, __c)} {}
+
+  _LIBCPP_HIDE_FROM_ABI zoned_time(string_view __name, const local_time<_Duration>& __tp, choose __c)
+    requires(requires {
```
Comment (ldionne) [inline libcxx/include/__chrono/zoned_time.h:118]: Constraint.

File: libcxx/include/__chrono/zoned_time.h [hunk 4]
```diff
@@ -76,12 +81,71 @@ class zoned_time {
...
+
+  // per wording choose has no effect
+  template <class _Duration2, class _TimeZonePtr2>
+  _LIBCPP_HIDE_FROM_ABI zoned_time(_TimeZonePtr __zone, const zoned_time<_Duration2, _TimeZonePtr2>& __zt, choose)
+    requires is_convertible_v<sys_time<_Duration2>, sys_time<_Duration>>
+      : __zone_{std::move(__zone)}, __tp_{__zt.get_sys_time()} {}
+
+  template <class _Duration2, class _TimeZonePtr2>
+  _LIBCPP_HIDE_FROM_ABI zoned_time(string_view __name, const zoned_time<_Duration2, _TimeZonePtr2>& __zt)
+    requires(requires {
```
Comment (ldionne) [inline libcxx/include/__chrono/zoned_time.h:137]: Constraint.

File: libcxx/include/__chrono/zoned_time.h [hunk 5]
```diff
@@ -76,12 +81,71 @@ class zoned_time {
...
+      _TimeZonePtr{__traits::locate_zone(string_view{})};
+    } && is_convertible_v<sys_time<_Duration2>, sys_time<_Duration>>)
+      : zoned_time{__traits::locate_zone(__name), __zt} {}
+
+  template <class _Duration2, class _TimeZonePtr2>
+  _LIBCPP_HIDE_FROM_ABI zoned_time(string_view __name, const zoned_time<_Duration2, _TimeZonePtr2>& __zt, choose __c)
+    requires(requires {
```
Comment (ldionne) [inline libcxx/include/__chrono/zoned_time.h:144]: Constraint.

File: libcxx/test/std/time/time.zone/time.zone.zonedtime/time.zone.zonedtime.ctor/string_view_local_time_choose.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <chrono>
+#include <concepts>
+#include <cassert>
+
+#include "test_offset_time_zone.h"
+
+int main(int, char**) {
+  // Tests unique conversions. To make sure the test is does not depend on changes
+  // in the database it uses a time zone with a fixed offset.
```
Comment (ldionne) [inline libcxx/test/std/time/time.zone/time.zone.zonedtime/time.zone.zonedtime.ctor/string_view_local_time_choose.pass.cpp:30]:
```suggestion
  // Tests unique conversions. To make sure the test does not depend on changes
  // in the database it uses a time zone with a fixed offset.
```

File: libcxx/test/std/time/time.zone/time.zone.zonedtime/time.zone.zonedtime.ctor/string_view_local_time_choose.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    // 1 DE CE%sT 1980
+    // 1 E CE%sT
+    //
+    // ...
+    // R E 1981 ma - Mar lastSu 1u 1 S
+    // R E 1996 ma - O lastSu 1u 0 -
+
+    using namespace std::literals::chrono_literals;
+    {
+      std::chrono::zoned_time<std::chrono::seconds> zt{
```
Comment (ldionne) [inline libcxx/test/std/time/time.zone/time.zone.zonedtime/time.zone.zonedtime.ctor/string_view_local_time_choose.pass.cpp:52]: You could consider using `namespace cr = std::chrono` to shorten these tests.

**Review by ldionne — CHANGES_REQUESTED** [review]
This generally LGTM but the constraints seem wrong, so I'd like to see again once that is figured out.

**Comment by h-vetinari** [comment]
Peanut gallery comment: it would be very cool if `zoned_time` could make it for libcxx 19; 🤞 
It's IMO one of the most useful classes in `<chrono>`.

File: libcxx/include/__chrono/zoned_time.h (hunk 1 above)
Comment (mordante) [reply libcxx/include/__chrono/zoned_time.h:107]: See the comment about this in the parent patch. I'll mark the similar comments as resolved.

**Review by mordante — COMMENTED** [review]
> Peanut gallery comment: it would be very cool if `zoned_time` could make it for libcxx 19; 🤞 It's IMO one of the most useful classes in `<chrono>`.

I hope to get `zoned_time` and the missing clocks ready before LLVM-19. However it will be closer to the deadline than I hoped.

**Review by ldionne — APPROVED** [review]

**Pull request merged**